### ISSUE TITLE
Suggestion for general and combat tab setup

### ISSUE THREAD

**Issue by dazdya** [nihlaeth/dnd#13]
General tab
Race: choicebox as it is now.
Class: a button with generated text such as (fighter2/ wizard 1) or similar. Button opens a dialog.
    Line with:
    "lvl" "class (choicebox)" "hp (dieroll)"
    "1" "fighter (which is starting class)" "(which is max dieroll)"
    "2" "fighter" "7 (what I rolled on the d10)"
    "3" "" "wizard" "2 ( result of the d6)"
    next line has a button: "new level", which adds 4th level.
Next are the abilities,
"Strength" ""(buttonwith value)" "(modifier)" "current value in red or green, but only if different from normal"
Rest of the abilities is the same.

It seems best to record ability damage, instead of current value, just like with hit points. We just show the current value.
Afterthought: just like with the spells, this tab could use a rest button. Can a button influence multiple tabs easily?

    Combat tab
This is mostly derived values.
Hit points are calculated from the level die rolls and the constitution modifier.
AC comes from the dex modifier and the equipment, and maybe magic.
To hit is from classes and strength, and from weapon skills.
Weapon statistics come from equipment.
Saves from classes and abilities again, and maybe equipment.
Only damage needs to be a button. I think.

It's still important to make it a separate tab, because combat is so specific.

    Other tab suggestions will follow. Please let me know what you think, if I am being clear.

**Comment by nihlaeth** [comment]
I agree with your idea for the class interface, there is only one little thing:

>  "lvl" "class (choicebox)" "hp (dieroll)"

should be:
```
"lvl" "class (choicebox)" "hp (dieroll + [con])"
```
Because we don't record ability history, and I don't think we should start.

>  next line has a button: "new level", which adds 4th level.

We calculate level from experience, so this is not necessary. 

> It seems best to record ability damage, instead of current value, just like with hit points. We just show the current value.

Right now we have the base value, the temporary adjustment (damage, but also temporary positive changes), and the ability points you can assign every four levels. I cannot calculate the ability from only the damage. How do you see this form?

> Afterthought: just like with the spells, this tab could use a rest button. Can a button influence multiple tabs easily?

There's nothing that limits how many tabs a button can influence. I'm planning on adding a separate tab for actions, I only parked the rest button with spells because there was not a better place for it yet.

I completely agree with the combat tab.

**Comment by nihlaeth** [comment]
By the way, are you sure you want to move the abilities to the general tab? I gave them their own tab because they take up a moderate amount of space, you can collapse them separately this way.

**Comment by dazdya** [comment]
- We don't want to include the con-modifier. If the ability changes, we want to calculate the hit points in the combat tab, we don't want to have to go back and change all the die rolls. It's easiest to keep your entry data as clean as possible.

- Later, if and when campaigns are added, we can get level from the xp that the dm gives, but let's first get a working character sheet. And that means, for now, letting the player set the level himself. Simplest solutions first.

- You have a point with the ability form. Let's think about what can happen:
1/ the original die roll is rolled by the player.
2/ race modifiers happen.
3/ ancestor modifiers happen.
4/ every fourth level we get a +1.
5/ prayers and spells can change abilities. Some changes are sustainable, making them semi-permanent (for instance, the majesty spell, +1 charisma per circle).
6/ philtres from herbalism give a +3, but only temporary. Likewise, cooking, performances and ritual of fertility grants a temporary bonus. Maybe even more skills.
7/ warlock powers drain the caster's abilities.
8/ Some monsters have draining effects.
9/ resting removes one damage from every ability.
10/ and then there's even some powerful magic that will permanently change stuff, such as being raised from the dead.

1, to 4 only happen once. They can be taken into the standard ability score. 5 and 6 are positive, 7 and 8 negative. 9 (resting) removes the negative, but not the positive. And number 10 is so rare, it may be best to do that manually.
What if we calculate the ability that we show on the sheet using only 1 through 4, and we have 2 variables besides it, a positive modifier for 5 and 6, and a negative modifier for 7 and 8? Those would generate a current ability score (button), and clicking the button would show the modifiers. We could clone the interface from the spells there, so that if you stop maintaining your majesty, your charisma drops down, and you can also choose to end it while looking at your ability score. And we could use the same interface for philtres and skill bonuses, so we just click the x when the effect has ended. Damage can be healed by resting, or by other magic, such as restoration, which can be done manually.
I think this would give us a framework to place most actions people can do. I think. How does it sound to you?

- In placing the abilities on the general tab, I think about what activities go together. Abilities sound like the general info you keep at the top of your sheet. Other tabs should imply other actions, such as for skills and combat.

**Comment by nihlaeth** [comment]
> We don't want to include the con-modifier. If the ability changes, we want to calculate the hit points in the combat tab, we don't want to have to go back and change all the die rolls. It's easiest to keep your entry data as clean as possible.

I think I got myself confused as to how Max hit points are calculated. Which is it?
1. `sum(level_up_die_rolls) + current_constitution_modifier`
2. `sum(level_up_die_rolls) + level * current_constitution_modifier`
3. `sum(level_up_die_rolls) + sum(constitution_modifiers_at_time_of_level_up)`

> Later, if and when campaigns are added, we can get level from the xp that the dm gives, but let's first get a working character sheet. And that means, for now, letting the player set the level himself. Simplest solutions first.

The XP calculation is already in place and I feel very little for removing it only to re-add it later. I agree that the dungeon master should be able to assign XP later on, but the user also should still be able to, since they might be keeping a character sheet for a campaign for which the dungeon master doesn't use our tools.

## abilities
Number two is already included automatically, only this is not displayed anywhere, causing confusion.

But apparently there are a lot of temporary effects on ability score that all work differently, and I find it hard to think up an interface that won't be very confusing. I'm already not happy about the current one.

First off,  everything that can be calculated automatically, should be, but it should be displayed somewhere. Maybe we should calculate the semi-permanent and temporary effects on the ability score from all the individual effects put together. Maybe form rows that look like this:
`[amount] [type of effect](drop-down) [comment describing effect] <x>`
With a button for adding new rows?

Spells with target caster aren't hard to include automatically, but I don't don't think we should do that when the target isn't the same character. I don't want players influencing other character sheets... That's getting too close to being an online game, and it's sensitive to abuse.

> In placing the abilities on the general tab, I think about what activities go together. Abilities sound like the general info you keep at the top of your sheet. Other tabs should imply other actions, such as for skills and combat.

I tried to keep the tabs small, developing with small screens in mind. I figured since modifiers and ability scores are frequently used information, they should have their own tab. But that's only true for paper character sheets, since I think we will barely use that info directly when everything is calculated and displayed in the combat tab. I don't feel strongly either way, so we'll do it your way.

**Comment by dazdya** [comment]
The correct hp formula is this, 
> sum(level_up_die_rolls) + level * current_constitution_modifier
BUT
A level 1 character gets hp as if he rolled the maximum on the die roll, and on level 11 and higher, the character get just 1 hp / level, without [con].

**Comment by nihlaeth** [comment]
That's good to know, I will work that in. And with this formula I absolutely agree that we should not include constitution modifier in the die roll field.

**Comment by dazdya** [comment]
About the abilities, I think we agree. By the way, I forgot age modifiers. We should put those in somewhere.

**Comment by nihlaeth** [comment]
# to do
- [x] merge general and abilities tab
- [x] refactor class interface to be level based
- [ ] add age to general tab
- [ ] add background to general tab
- [ ] refactor abilities interface for improved temporary and semi-permanent effects
- [x] add combat tab
- [ ] add action tab with rest button
- [x] automate calculating max HP
- [x] move HP to combat tab
- [x] move available spell slots to be underneath prepared spells
- [x] whitespace between spell circles
- [x] display circle behind spell name
- [x] indicate error when preparing spells in circles character doesn't have access to
- [x] indicate circles with roman numerals

I will add to this list when you get around to looking at the spell interface
